presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/BaseCostBasedPlanTest.java

```diff
@@ -0,0 +1,248 @@
+import com.facebook.presto.sql.planner.Plan;
+import com.facebook.presto.sql.planner.SimplePlanVisitor;
+import com.facebook.presto.sql.planner.assertions.BasePlanTest;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
+import com.facebook.presto.sql.planner.plan.TableScanNode;
+import com.facebook.presto.sql.planner.plan.ValuesNode;
+import com.google.common.base.Strings;
+import com.google.common.io.Files;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.stream.Stream;
+
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.REPLICATED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.testing.TestngUtils.toDataProvider;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.nio.file.Files.exists;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static org.testng.Assert.assertEquals;
+
+public abstract class BaseCostBasedPlanTest
+        extends BasePlanTest
+{
+    public BaseCostBasedPlanTest(LocalQueryRunnerSupplier supplier)
+    {
+        super(supplier);
+    }
+
+    protected abstract Stream<Query> getQueries();
+
+    protected abstract Path getExpectedJoinOrderingFile(String queryId);
+
+    @DataProvider
+    public Object[][] getQueriesDataProvider()
+    {
+        return getQueries()
+                .collect(toDataProvider());
+    }
+
+    @Test(dataProvider = "getQueriesDataProvider")
+    public void test(Query query)
+    {
+        assertEquals(generateQueryPlan(query.getQuery()), read(resolve(getExpectedJoinOrderingFile(query.getId()))));
+    }
+
+    private Path resolve(Path path)
+    {
+        // Test is run with module as current working directory. generate() is run with project's top-level directory as current working directory.
+        if (exists(path)) {
+            return path;
+        }
+        Path fromParent = Paths.get("..").resolve(path);
+        checkState(exists(fromParent), "Unable to resolve: " + path);
+        return fromParent;
+    }
+
+    public void generate()
+            throws Exception
+    {
+        initPlanTest();
+        try {
+            getQueries()
+                    .parallel()
+                    .forEach(query -> {
+                        try {
+                            Files.write(generateQueryPlan(query.getQuery()).getBytes(UTF_8), getExpectedJoinOrderingFile(query.id).toFile());
```

Comment:
Instead of using `getExpectedJoinOrderingFile` i would suggest adding a `getPrestoMainSourceRoot` method, and use it to get the source root. In this method you can explicitly check that the home directory, from which this class is being executed is within the `presto-main`, and display a meaningful error message saying that this class must be executed from so. Append the `src/test/resources`, and concatenate with the `expectedPlanResourcePath` from the `Query` object.

Reply:
There is one downside of this approach, which so far I wanted to avoid. Notice that now tpcds tests are using resources from `presto-main` (previously it was using `presto-tpcds`). Maybe it is not a big deal, because queries are taken from `presto-benchto-queries` anyway.